src/NHapi.Base/Model/AbstractGroup.cs

```diff
@@ -1,485 +1,559 @@
+    /// groups of segments (and/or other sub-groups) that are implicitly defined by message structures
+    /// in the HL7 specification.  A subclass should define it's group structure by putting repeated calls to
+    /// the add(...) method in it's constructor.  Each call to add(...) adds a specific component to the
+    /// Group.
+    /// </summary>
+    /// <author>  Bryan Tripp (bryan_tripp@sourceforge.net).
+    /// </author>
+    public abstract class AbstractGroup : IGroup
+    {
+        private static readonly IHapiLog Log;
+
+        private List<AbstractGroupItem> items;
+        private IGroup parentStructure;
+        private IModelClassFactory myFactory;
+
+        static AbstractGroup()
+        {
+            Log = HapiLogFactory.GetHapiLog(typeof(AbstractGroup));
+        }
+
+        /// <summary> This constructor should be used by implementing classes that do not
+        /// also implement Message.
+        ///
+        /// </summary>
+        /// <param name="parentStructure">the group to which this Group belongs.
+        /// </param>
+        /// <param name="factory">the factory for classes of segments, groups, and datatypes under this group.
+        /// </param>
+        protected internal AbstractGroup(IGroup parentStructure, IModelClassFactory factory)
+        {
+            this.parentStructure = parentStructure;
+            myFactory = factory;
+            Init();
+        }
+
+        /// <summary> This constructor should only be used by classes that implement Message directly.
+        ///
+        /// </summary>
+        /// <param name="factory">the factory for classes of segments, groups, and datatypes under this group.
+        /// </param>
+        protected internal AbstractGroup(IModelClassFactory factory)
+        {
+            myFactory = factory;
+            Init();
+        }
+
+        /// <summary>
+        /// A string[] of group names.
+        /// </summary>
+        public virtual string[] Names
+        {
+            get
+            {
+                string[] retVal = new string[items.Count];
+                for (int i = 0; i < items.Count; i++)
+                {
+                    AbstractGroupItem item = items[i];
+                    retVal[i] = item.Name;
+                }
+
+                return retVal;
+            }
+        }
+
+        /// <summary> Returns the Message to which this segment belongs.</summary>
+        public virtual IMessage Message
+        {
+            get
+            {
+                IStructure s = this;
+                while (!typeof(IMessage).IsAssignableFrom(s.GetType()))
+                {
+                    s = s.ParentStructure;
+                }
+
+                return (IMessage)s;
+            }
+        }
+
+        /// <summary>Returns the parent group within which this structure exists (may be root
+        /// message group).
+        /// </summary>
+        public virtual IGroup ParentStructure

```

Comment:
Can we get rid of the private field and just use the property?

Reply:
done